File: code/python/剑指offer12.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def backtrack(index,length,i,j,row,col,used):
            if index==length:
                return True
            direction = [(-1,0),(1,0),(0,-1),(0,1)]
            res = False
            for d in direction:
                if i+d[0]>=0 and i+d[0]<row and j+d[1]>=0 and j+d[1]<col:
                    if board[i+d[0]][j+d[1]]==word[index] and used[i+d[0]][j+d[1]]==False:
                        used[i+d[0]][j+d[1]] = True
                        res = res or backtrack(index+1,length,i+d[0],j+d[1],row,col,used)
                        used[i+d[0]][j+d[1]] = False
            return res

        row = len(board)
        col = len(board[0])
        length = len(word)
        for i in range(row):
            for j in range(col):
                if board[i][j]==word[0]:
                    used = [[False]*col for _ in range(row)]
                    used[i][j] = True
                    if backtrack(1,length,i,j,row,col,used)==True:
                        return True

        return False

File: code/java/剑指offer12.java
class Solution {
    public boolean backtrack(String word,char[][] board,int index,int length,int i,int j,int row,int col,boolean[][] used){
        if(index==length) return true;
        int[] d_r = {-1,1,0,0};
        int[] d_c = {0,0,-1,1};
        boolean res = false;
        for(int d=0;d<4;d++){
            if(i+d_r[d]>=0&&i+d_r[d]<row&&j+d_c[d]>=0&&j+d_c[d]<col){
                if(!used[i+d_r[d]][j+d_c[d]]&&word.charAt(index)==board[i+d_r[d]][j+d_c[d]]){
                    used[i+d_r[d]][j+d_c[d]] = true;
                    res = res|backtrack(word,board,index+1,length,i+d_r[d],j+d_c[d],row,col,used);
                    used[i+d_r[d]][j+d_c[d]] = false;
                }
            }
        }
        return res;
    }

    public boolean exist(char[][] board, String word) {
        int row = board.length;
        int col = board[0].length;
        int length = word.length();
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                if(board[i][j]==word.charAt(0)){
                    boolean[][] used = new boolean[row][col];
                    used[i][j] = true;
                    if(backtrack(word,board,1,length,i,j,row,col,used)) return true;
                }
            }
        }
        return false;
    }
}